dg.go

```diff
@@ -265,14 +265,14 @@ func (n *node[NodeType]) Item() NodeType {
 }
 
 func (n *node[NodeType]) ResolutionStatus() ResolutionStatus {
-	// The status can change, so lock to prevent data races.
+	// The status can change and be accessed on different threads, so lock to prevent races.
 	n.dg.lock.Lock()
 	defer n.dg.lock.Unlock()
 	return n.status
 }
 
 func (n *node[NodeType]) IsReady() bool {
-	// The ready state can change, so lock to prevent data races.
+	// The status can change and be accessed on different threads, so lock to prevent races.
```

Comment:
I suppose we could set up a more involved hypothetical where the lock would actually help: and that's because "the only viable reason" to call `IsReady` would be because you expect to read the node's dependencies knowing that the node is now ready. Indirectly (and very inefficiently) the lock allows this by providing a memory barrier so that, having seen the `IsReady() == true`, you know that you can't subsequently read older values from other node properties.

Using `sync/atomic` `Load` variants would, as I've said, be a clearer way to do this; but, given the general level of coarse lock-based synchronization within ARCA, it'd make a lot more sense to move the lock out of this function and hold it around the equivalent the `if node.ready { ... }` rather than using atomics or using a detached lock as an expensive sledgehammer-style memory barrier.

Reply:
I adjusted the comments. If you want a new design, let's have a call on Monday.

Reply:
As [I said elsewhere](https://github.com/arcalot/go-dgraph/pull/15#pullrequestreview-2105051432), I'm OK with the present code, so long as it's used only in testing and not in production operation.

The new design that I think we need is for the Engine's deadlock detection.  Given that this PR is in support of the Engine, it wouldn't be out of scope to add the support for that design in this PR.  However, it wouldn't be unreasonable to consider that separately, either, but I think it's going to gate the Engine work.